### ISSUE TITLE
Reduce not needed traffic in case of WebDriver Classic serialization via BiDi

### ISSUE THREAD

**Issue by sadym-chromium** [w3c/webdriver-bidi#390]
To implement the WebDriverClassic serialization, the user needs to serialize all JSON-serializable objects to the very end, so they will have to use `MaxDepth=MAX_INT`. But with current BiDi serialization it can lead to a HUGE response, if there will be some nodes close to root, as it will end up with serialization of a big part of the DOM tree, which is probably not what the user really wants.

We suggest either:
* Remove field `children` from the `NodeRemoteValue` completely, and rely on the user to construct in-depth serialization if the really need it. Or
* Add additional parameter `MaxDomDepth` to limit the response only to the information required for constructing WebDriver Classic serialized values.

**Comment by whimboo** [comment]
Hm, I'm not sure that I understand this request. In WebDriver classic elements and shadow roots are serialized as a WebElement reference and only contains the field `element-6066-11e4-a52e-4f735466cecf` or `shadow-6066-11e4-a52e-4f735466cecf` each with a unique id. There are no other fields set.

@sadym-chromium maybe you can explain on a given example what's problematic?

**Comment by sadym-chromium** [comment]
@whimboo let's say the user wants to get result of the following line via WebDriver Classic:
```
{
  some_deep_object: {
    b: 42, 
    foo: 'bar'
  }, 
  document: document
}
```

The expected result will be
```
{
    "value": {
        "some_deep_object": {
            "b": 42,
            "foo": "bar"
        },
        "document": {
            "element-6066-11e4-a52e-4f735466cecf": "25F87A2F0A0535662786F1925D14AF05_element_2"
        }
    }
}```

In order to be able to construct this result based on the BiDi response, the user has to pass argument `MaxDepht=MAX_INT`, as long as they don't know how deep `some_deep_object` can be. But the BiDi response will contain all the DOM nodes serialized, as long as `MaxDepth` is used both for DOM and for primitive serialization.

```
{
    "result": {
        "type": "object",
        "value": [
            [
                "a",
                {
                    "type": "object",
                    "value": [
                        [
                            "b",
                            {
                                "type": "number",
                                "value": 42
                            }
                        ],
                        [
                            "foo",
                            {
                                "type": "string",
                                "value": "bar"
                            }
                        ]
                    ]
                }
            ],
            [
                "document",
                {
                    "type": "node",
                    "value": {
                        "nodeType": 9,
                        "sharedId": "SOME_SHARED_ID",
                        "children": [ALL_THE_DOM_HERE]
                    }
                }
            ]
        ]
    }
}
```

**Comment by jgraham** [comment]
You can just set `maxDepth=null` rather than `MAX_INT`, I believe.

Anyway, I agree that it makes sense to have a different limit for DOM nodes compared to JS objects. I wonder if we should create a `script.SerializationOptions` type with different fields, rather than putting every field at the top level for each object. That might also be better for cases where we may need session-level configuration (e.g. for setting the behaviour for events).

**Comment by whimboo** [comment]
> @whimboo let's say the user wants to get result of the following line via WebDriver Classic:
> 
> ```
> {
>   some_deep_object: {
>     b: 42, 
>     foo: 'bar'
>   }, 
>   document: document
> }
> ```
> 
> The expected result will be
> 
> ```
> {
>     "value": {
>         "some_deep_object": {
>             "b": 42,
>             "foo": "bar"
>         },
>         "document": {
>             "element-6066-11e4-a52e-4f735466cecf": "25F87A2F0A0535662786F1925D14AF05_element_2"
>         }
>     }
> }
> ```

The expected result here will actually be `javascript error` when a script tries to return such an object via `execute (async) script`. The reason is that `document` is not a valid `Element` or `Node`, and given that `document` is cyclic  we are not able to serialize this data in WebDriver classic - at least with the current status of the specification. But I believe someone could have the same example with a real element or shadow root.

Otherwise what James said and I agree that an option struct would make sense. I've actually introduced that already in our implementation for Firefox but only to pass-down the `node cache` to the appropriate helper methods.

**Comment by sadym-chromium** [comment]
> The expected result here will actually be `javascript error` when a script tries to return such an object via `execute (async) script`. The reason is that `document` is not a valid `Element` or `Node`, and given that `document` is cyclic we are not able to serialize this data in WebDriver classic - at least with the current status of the specification. But I believe someone could have the same example with a real element or shadow root.

Hm. I read this as `document` IS a node: https://dom.spec.whatwg.org/#interface-document.

**Comment by whimboo** [comment]
Sorry my fault. I actually wanted to say `is not a valid Element or ShadowRoot`.

**Comment by whimboo** [comment]
@sadym-chromium can we now close this issue as fixed given that https://github.com/w3c/webdriver-bidi/pull/385 has been merged?

**Comment by sadym-chromium** [comment]
> @sadym-chromium can we now close this issue as fixed given that #385 has been merged?

yes, it is addressed
